base/iterators.jl

```diff
@@ -649,6 +649,52 @@ end
 iterate(it::Drop, state) = iterate(it.xs, state)
 isdone(it::Drop, state) = isdone(it.xs, state)
 
+
+# takewhile
+
+takewhile(pred::Function,xs) = TakeWhile(xs,pred)
+struct TakeWhile{I}
+    xs::I
+    pred
+end
+
+iterate(ibl::TakeWhile{I},itr...) where {I} = begin
+    y = iterate(ibl.xs,itr...)
+    y === nothing && return nothing
+    ibl.pred(y[1]) || return nothing
+    y
+end
+
+IteratorSize(ibl::TakeWhile{I}) where {I} = SizeUnknown()
+eltype(::Type{TakeWhile{I}}) where {I} = eltype(I)
+IteratorEltype(::Type{TakeWhile{I}}) where {I} = IteratorEltype(I)
+
+
+# dropwhile
+
+dropwhile(pred::Function,itr) = DropWhile(itr,pred)
+mutable struct DropWhile{I}
+    xs::I
+    pred
+    fltd::Bool
```

Comment:
This should be part of the iteration state instead of making the struct mutable.